go/chat/inboxsource.go

```diff
@@ -685,6 +691,132 @@ func (s *HybridInboxSource) ReadUnverified(ctx context.Context, uid gregor1.UID,
 	return res, err
 }
 
+type nameContainsQueryRes int
+
+const (
+	nameContainsQueryNone nameContainsQueryRes = iota
+	nameContainsQuerySimilar
+	nameContainsQueryPrefix
+	nameContainsQueryExact
+)
+
+type convSearchHit struct {
+	conv      types.RemoteConversation
+	queryToks []string
+	convToks  []string
+	hits      []nameContainsQueryRes
+}
+
+func (h convSearchHit) hitScore() (score int) {
+	exacts := 0
+	for _, hit := range h.hits {
+		switch hit {
+		case nameContainsQueryExact:
+			score += 20
+			exacts++
+		case nameContainsQueryPrefix:
+			score += 10
+		case nameContainsQuerySimilar:
+			score += 3
+		}
+	}
+	if len(h.queryToks) == len(h.convToks) && len(h.hits) == len(h.convToks) {
+		if exacts == len(h.hits) {
+			return 1000000
+		}
+	}
+	return score
+}
+
+func (h convSearchHit) less(o convSearchHit) bool {
+	hScore := h.hitScore()
+	oScore := o.hitScore()
+	if hScore < oScore {
+		return true
+	} else if hScore > oScore {
+		return false
+	}
+	return h.conv.GetMtime().Before(o.conv.GetMtime())
+}
+
+func (h convSearchHit) valid() bool {
+	return len(h.hits) > 0
+}
+
+func (s *HybridInboxSource) isConvSearchHit(conv types.RemoteConversation, queryToks []string,
+	username string) (res convSearchHit) {
+	var convToks []string
+	res.conv = conv
+	res.queryToks = queryToks
+	switch conv.GetMembersType() {
+	case chat1.ConversationMembersType_TEAM:
+		convToks = []string{utils.SearchableRemoteConversationName(conv, username)}
```

Comment:
move `utils.SearchableRemoteConversationName(conv, username)` since both branches use the value?